internal/clients/spotify/spotify.go

```diff
@@ -0,0 +1,246 @@
+package spotify
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/SongStitch/song-stitch/internal/constants"
+	"github.com/SongStitch/song-stitch/internal/models"
+	"github.com/rs/zerolog"
+)
+
+type SpotifyAuthResponse struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int    `json:"expires_in"` // seconds
+}
+
+type Token struct {
+	AccessToken string
+	ExpiresIn   int
+	client      *http.Client
+	endpoint    string
+}
+
+func (t *Token) Refresh() error {
+	client_id := os.Getenv("SPOTIFY_CLIENT_ID")
+	client_secret := os.Getenv("SPOTIFY_CLIENT_SECRET")
+
+	if client_id == "" || client_secret == "" {
+		return errors.New("spotify credentials not set")
+	}
+
+	data := url.Values{}
+	data.Set("grant_type", "client_credentials")
+	data.Set("client_id", client_id)
+	data.Set("client_secret", client_secret)
+
+	req, err := http.NewRequest(http.MethodPost, t.endpoint, strings.NewReader(data.Encode()))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	res, err := t.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return errors.New("spotify authentication failed")
+	}
+
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return err
+	}
+
+	var response SpotifyAuthResponse
+	err = json.Unmarshal([]byte(body), &response)
+	if err != nil {
+		return err
+	}
+	t.AccessToken = response.AccessToken
+	t.ExpiresIn = response.ExpiresIn
+	return nil
+}
+
+func (t *Token) KeepAlive(log zerolog.Logger) {
+	for {
+		log.Info().Msg("refreshing spotify token...")
+		err := t.Refresh()
+		if err != nil {
+			log.Error().Err(err).Msg("failed to refresh spotify token")
+		}
+		// Wait until 5 minutes before expiration and then refresh again
```

Comment:
Why 5 minutes, don't they expire every hour? We can maybe get away with a longer time.

Though 5min may make more sense, especially since the servers get stopped on fly.io. Say a server gets paused for 2hrs, but internally the sleep only ran for less than 5 mins, it would be expired.

Maybe it'd be better to also parse `ExpiresIn` to a timestamp from when it's grabbed and do a timestamp diff with `now()` to ensure a expired token is used?

Reply:
Oh sorry I misunderstood (thought it was every 5 mins). I think we're cutting it close. If the server stops for more than 5 minutes AFAIK `time.sleep` would stop too and it'll go over 1hr before doing another refresh. Perhaps drop it to 40 mins or so, just to add a bit of a buffer?

I think we should be fine, fly.io sends a `SIGINT` signal, then does a resume. Looking at the logs, looks like it reloads the `.env` file on startup again (though I'm not sure if it's passing in a `SIGCONT`), which leads me to believe it would get a new token upon resuming. Good thing you're logging token refreshes—if we see it refresh upon a restart we should be good

<img width="1212" alt="image" src="https://github.com/SongStitch/song-stitch/assets/856001/cd6f5412-f9c0-4764-9d17-a4c677dded39">

Reply:
As far as I know, when the pod gets restarted, it restarts the whole application (you can see in the logs there "Error loading .env file" and "Starting server...") so it should refresh the token correctly.

Reply:
Yeah I think you're right, logs say SIGINT initially, but it exits with signal 15 (`SIGTERM`), so agree I think we're safe here